packages/vscode-extension/src/debug/depsChecker/funcToolChecker.ts

```diff
@@ -51,187 +46,210 @@ export class FuncToolChecker implements IDepsChecker {
     await this._logger.info(
-      Messages.startInstallFunctionCoreTool.replace("@NameVersion", installedNameWithVersion)
+      Messages.finishInstallFunctionCoreTool.replace("@NameVersion", displayFuncName)
+    );
+  }
+
+  private async handleInstallFuncFailed(): Promise<void> {
+    await this.cleanup();
+
+    this._telemetry.sendSystemErrorEvent(
+      DepsCheckerEvent.funcInstallError,
+      TelemtryMessages.failedToInstallFunc,
+      Messages.failToValidateFuncCoreTool.replace("@NameVersion", displayFuncName)
+    );
+
+    throw new DepsCheckerError(
+      Messages.failToInstallFuncCoreTool.replace("@NameVersion", displayFuncName),
+      defaultHelpLink
     );
+  }
 
+  private async validate(): Promise<boolean> {
+    let isInstallationValid = false;
     try {
-      await this._telemetry.sendEventWithDuration(
-        DepsCheckerEvent.funcInstallCompleted,
-        async () => {
-          await this._adapter.runWithProgressIndicator(async () => {
-            await this.installFuncCoreTools(FuncVersion.v3);
-          });
-        }
-      );
-    } catch (error) {
+      const portableFunc = await this.queryFuncVersion(FuncToolChecker.getPortableFuncExecPath());
+      isInstallationValid = portableFunc !== null && supportedVersions.includes(portableFunc);
+    } catch (err) {
       this._telemetry.sendSystemErrorEvent(
-        DepsCheckerEvent.funcInstall,
-        TelemtryMessages.failedToInstallFunc,
-        error
+        DepsCheckerEvent.funcValidationError,
+        TelemtryMessages.failedToValidateFunc,
+        err
       );
+    }
 
-      throw new DepsCheckerError(
-        Messages.failToInstallFuncCoreTool.replace("@NameVersion", installedNameWithVersion),
-        defaultHelpLink
-      );
+    if (!isInstallationValid) {
+      this._telemetry.sendEvent(DepsCheckerEvent.funcValidationError);
     }
+    return isInstallationValid;
+  }
 
-    // validate after installation.
-    const isInstalled = await this.isInstalled();
-    if (!isInstalled) {
-      this._telemetry.sendSystemErrorEvent(
-        DepsCheckerEvent.funcValidation,
-        TelemtryMessages.failedToInstallFunc,
-        Messages.failToValidateFuncCoreTool.replace("@NameVersion", installedNameWithVersion)
-      );
+  private handleNpmNotFound() {
+    this._telemetry.sendEvent(DepsCheckerEvent.npmNotFound);
+    throw new DepsCheckerError(
+      Messages.needInstallFuncCoreTool.replace("@NameVersion", displayFuncName),
+      defaultHelpLink
+    );
+  }
 
-      throw new DepsCheckerError(
-        Messages.failToInstallFuncCoreTool.replace("@NameVersion", installedNameWithVersion),
-        defaultHelpLink
-      );
+  private static getDefaultInstallPath(): string {
+    return path.join(os.homedir(), `.${ConfigFolderName}`, "bin", "func");
+  }
+
+  private static getPortableFuncExecPath(): string {
+    return path.join(FuncToolChecker.getDefaultInstallPath(), "func");
+  }
+
+  public async getFuncExecPath(): Promise<string> {
```

Comment:
IMO, it is better to change this function to `async getFuncCommand(): Promise<string[]>`. 
- For portable version, return `[FuncToolChecker.getPortableFuncExecPath()]`;
- For global version, return something like `["npx", "func"]` (you may need to change the command according to @qinezh 's comment).

#### The reasons are:

1. The return value is not a path as the function name suggests. It is a command.
2. Although currently the code may work, if you escape the path in `createBackendStartTask`, this code will fail because the shell will search for the file named "npx func" (as a whole string), which does not exist.

Reply:
has talked offline